File: PracticasPDOOJava/Practica2_Java/src/irrgarten/Game.java
package irrgarten;

import java.util.ArrayList;

/**
 * Núcleo del juego: gestiona jugadores, monstruos, el laberinto, el turno 
 * y el registro de eventos (log).
 * @author Joaquín Cruz Lorenzo
 */
public class Game {
    // Número máximo de rondas
    private static final int MAX_ROUNDS = 10;
    
    private  int currentPlayerIndex;
    private String log;
    
    private final ArrayList<Player> players = new ArrayList<>();
    private final ArrayList<Monster> monsters = new ArrayList<>();
    private final Labyrinth labyrinth;
    
    /**
     * Crea una partida con {@code nplayers} jugadores. Decide quién empieza, 
     * crea el laberinto y realiza una confifuración mínima (colocar salida y monstruos).
     * @param nplayers número de jugadores
     */
    public Game(int nplayers) {
        // 1) Crear jugadores (ejemplo: letras 'A', 'B', 'C', ...)
        for(int i = 0; i < nplayers; i++){
            char number = (char) ('A' + i);
            float intel = Dice.randomIntelligence();
            float str = Dice.randomStrength();
            players.add(new Player(number, intel, str));
        }
        
        // 2) Quien empieza
        currentPlayerIndex = Dice.whoStarts(nplayers);
        
        // 3) Crear laberinto (dimensiones y salida de ejemplo)
        int rows = 6, cols = 8, exitRow = 0, exitCol = cols - 1;
        labyrinth = new Labyrinth(rows, cols, exitRow, exitCol);
        
        // 4) Configurar laberinto (bloques/monstruos)
        configureLabyrinth();
        
        // 5) Repartir jugadores (en P3; aquí puede no estar implementado)
        try {
            labyrinth.spreadPlayers(players);
        } catch (UnsupportedOperationException ex) {
            // Nada aún, se completa en P3
        }
     }
    
    /**
     * 
     * @return {@code true} si el juego ha finalizado (hay ganador)
     */
    public boolean finished() {
        return labyrinth.haveAWinner(); // Delega en el laberinto.
    }
    
    /**
     * Fabrica un {@link GameState} con información textual del tablero, 
     * jugadores y monstruos, el jugador actual, si hay ganador y el log.
     * @return  estado observable del juego
     */
    public GameState getGameState() {
        // Laberinto como texto
        String labTxt = labyrinth.toString();
        
        // Listado de jugadores
        StringBuilder ps = new StringBuilder();
        for (Player p : players) ps.append(p).append('\n');
        
        // Listado de monstruos
        StringBuilder ms = new StringBuilder();
        for (Monster m : monsters) ms.append(m).append('\n');
        
        return new GameState(labTxt, ps.toString(), ms.toString(), currentPlayerIndex, finished(), log);
    }
    
    /**
     * Configura el laberinto añadiendo monstruos.
     * Los monstruos se guardan también en el contendor de esta clase.
     */
     void configureLabyrinth() {
        // Crear algunos monstruos con valores aleatorios razonables
        for (int i = 0; i < 3; i++) {
            Monster m = new Monster("M" + i, Dice.randomIntelligence(), Dice.randomStrength());
            
            // Colocar una posición vacía aleatoria
            int[] pos = labyrinth.randomEmptyPos();
            labyrinth.addMonster(pos[Labyrinth.ROW], pos[Labyrinth.COL], m);
            monsters.add(m);
        }
    }
    
     /**
      * Pasa el turno al siguiente jugador.
      */
    void nextPlayer() {
        currentPlayerIndex = (currentPlayerIndex + 1) %players.size();
    }
    
    // ========================================
    // P3
    // ========================================
    public boolean nextStep(Directions preferredDirection) {
        throw new UnsupportedOperationException();
    }
    
    Directions actualDirection(Directions preferredDirection) {
        throw new UnsupportedOperationException();
    }
    
    GameCharacter combat(Monster monster) {
        throw new UnsupportedOperationException();
    }
    
    void manageReward(GameCharacter winner) {
        throw new UnsupportedOperationException();
    }
    
    void manageResurrection(){
        throw new UnsupportedOperationException();
    }
    
    
    // ========================================
    // MÉTODOS DE LOG 
    // ========================================
    
    void logPlayerWon() { log += "Jugador ha ganado!\n"; }
    void logMonsterWon() { log += "Monstruo ha ganado!\n"; }
    void logResurrected() { log += "Jugador resucitado.\n"; }
    void logPlayerSkipTurn() { log += "Jugador ha perdido turno (muerto)\n"; }
    void logPlayerNoOrders() { log += "Jugador no ha recibido ordenes validas\n"; }
    void lonNoMonster() { log += "No hay monstruo en esta casilla / sin movimiento\n"; }
    void logRounds(int rounds, int max) { log += "Rondas: " + rounds + " / " + max + "\n"; }
}

File: PracticasPDOORuby/Practica2_Ruby/lib/Irrgarten/game.rb
# frozen_string_literal: true
require_relative 'player'
require_relative 'monster'
require_relative 'labyrinth'
require_relative 'game_state'
require_relative 'dice'

module Irrgarten
  class Game
    # --- Constantes y Atributos de Lectura ---
    MAX_ROUNDS = 10

    # ===========================================================
    # MÉTODOS PÚBLICOS
    # ===========================================================

    # Constructor
    # @param nplayers Número de jugadores en la partida
    # @return Objeto de la clase Game
    def initialize(nplayers)
      @players = Array.new(nplayers) { |i| Player.new(i, Dice.random_intelligence, Dice.random_strength) }
      @monsters = []
      @log = ""
      @labyrinth = Labyrinth.new(10, 10, 0, 0)
      @current_player_index = Dice.who_starts(nplayers)

      # Llamada a un método privado para la configuración inicial
      configure_labyrinth

      # Distribuir jugadores en el laberinto
      # @labyrinth.spread_player(@players)
    end

    # Determina si la partida ha finalizado comprobando si hay un ganador en el laberinto.
    # @return [Boolean] true si hay un ganador, false en caso contrario
    def finished?
      @labyrinth.have_a_winner?
    end

    # Obtiene el estado actual del juego, incluyendo el laberinto, los jugadores, los monstruos,
    # el índice del jugador actual, si hay un ganador y el registro de eventos.
    # @return [GameState] Objeto que representa el estado actual del juego
    def get_game_state
      GameState.new(@labyrinth.to_s, @players.map(&:to_s).join("\n"), @monsters.map(&:to_s).join("\n"), @current_player_index, finished?, @log)
    end

    # ===========================================================
    # P3 - Métodos públicos a implementar
    # ==========================================================
    def next_step(preferred_direction) = raise NotImplementedError, "P3"



    # ===========================================================
    # MÉTODOS PRIVADOS
    # ===========================================================
    private

    # Configura el laberinto inicial con monstruos y otras configuraciones necesarias.
    # Este método es llamado desde el constructor.
    # @return Nada (modifica el estado del laberinto y los monstruos)
    def configure_labyrinth
      # Añadir algunos monstruos de ejemplo
      5.times do |i|
        monster = Monster.new("Monster#{i}", Dice.random_intelligence, Dice.random_strength)
        pos = @labyrinth.random_empty_pos
        @labyrinth.add_monster(pos[Labyrinth::ROW], pos[Labyrinth::COL], monster)
        @monsters << monster
      end
    end

    # Actualiza el índice del jugador actual para pasar al siguiente jugador en la lista.
    # @return Nada (modifica el estado del juego)
    def next_player
      @current_player_index = (@current_player_index + 1) % @players.length
    end

    # Métodos de logging para registrar eventos importantes en el juego.
    def log_player_won
      @log += "El jugador ha ganado la partida!\n"
    end

    def log_monster_won
      @log += "El monstruo ha ganado la partid!\n"
    end

    def log_resurrected
      @log += "El jugador ha resucitado\n"
    end

    def log_player_skip_turn
      @log += "El jugador pierde el turno por estar muerto\n"
    end

    def log_player_no_orders
      @log += "El jugador no siguio las indicaciones\n"
    end

    def log_no_monster
      @log += "El jugador se ha movido a una casilla vacia o no le ha sido posible moverse\n"
    end

    def log_rounds(rounds, max)
      @log += "Ronda #{rounds} de #{max}\n"
    end

    # ===========================================================
    # P3 - Métodos privados a implementar
    # ===========================================================
    def actual_direction(preferred_direction) = raise NotImplementedError, "P3"
    def combat(monster) = raise NotImplementedError, "P3"
    def manage_reward(winner) = raise NotImplementedError, "P3"
    def manage_resurrection = raise NotImplementedError, "P3"
  end
end

